Testing of masked reductions on mixed layout inputs.

pearu opened the pull request:
The design of this feature is discussed in https://github.com/pytorch/pytorch/pull/71239#discussion_r787292751

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #72633
* #71239
* __->__ #72398
* #72397

Differential Revision: [D34408881](https://our.internmc.facebook.com/intern/diff/D34408881)

pearu committed:
Testing of masked reductions on mixed layout inputs.

[ghstack-poisoned]

cpuhrsch commented on test/test_masked.py:
```diff
@@ -162,6 +162,22 @@ def apply_masked_normalization_along_dim(op, input, *args, **kwargs):
+    """Convert the tensor content of object to strided tensor content.
+    """
+    if isinstance(obj, torch.Tensor):
+        if obj.layout == torch.strided:
```
I don't think this is needed. to_dense should check this and avoid conversion and if we're not sure we should check that it does.

pearu replied:
Ideally, I agree but we have this:
```python
>>> torch.tensor([[1, 2]]).to_dense()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NotImplementedError: Could not run 'aten::to_dense' with arguments from the 'CPU' backend
```

cpuhrsch replied:
Ok then we should add a stub under the [CompositeExplicitAutograd key](https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native#dispatch) in a separate PR.

pearu replied:
cref: https://github.com/pytorch/pytorch/issues/59958

cpuhrsch commented on test/test_masked.py:
```diff
@@ -162,6 +162,22 @@ def apply_masked_normalization_along_dim(op, input, *args, **kwargs):
+def to_strided(obj):
+    """Convert the tensor content of object to strided tensor content.
+    """
+    if isinstance(obj, torch.Tensor):
```
nit: You could also use [torch.is_tensor](https://pytorch.org/docs/master/generated/torch.is_tensor.html#torch.is_tensor).

pearu replied:
Fixed.

cpuhrsch commented on test/test_masked.py:
```diff
@@ -162,6 +162,22 @@ def apply_masked_normalization_along_dim(op, input, *args, **kwargs):
+            return obj
+        return obj.to_dense()
+    elif isinstance(obj, (tuple, list)):
+        return type(obj)(list(map(to_strided, obj)))
```
We have [pytree](https://github.com/pytorch/pytorch/blob/master/torch/utils/_pytree.py) in PyTorch that could simplify this (see an [example invocation](https://github.com/pytorch/pytorch/blob/c19cda5782f4c38140cb5c6aec08d5e44ec0c4cd/test/test_pytree.py#L131-L152))

pearu replied:
Thanks for the hint! Done.

mruberry commented on test/test_masked.py:
```diff
@@ -162,6 +162,22 @@ def apply_masked_normalization_along_dim(op, input, *args, **kwargs):
 
 masked_ops = [op for op in op_db if op.name.startswith('_masked.')]
 masked_ops_with_references = [op for op in masked_ops if op.name.rsplit('.', 1)[-1] in reference_functions]
+masked_ops_with_mixed_layouts_support = [op for op in masked_ops if op.supports_sparse or op.supports_sparse_csr]
```
Nice

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -539,8 +539,9 @@ def __init__(self,
-                 sample_inputs_func=None,  # function to generate sample inputs
+                 sample_inputs_func=None,  # function to generate sample inputs with strided-only layouts
                  error_inputs_func=None,  # function to generate inputs that will throw errors
+                 sample_inputs_mixed_func=None,  # function to generate sample inputs with mixed layouts
```
I would just add a sample input function for each layout

mruberry replied:
(if we want to enumerate all sample input functions we can have a list of all of them for tests to iterate over)

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1034,6 +1044,36 @@ def sample_inputs_masked_reduction(op_info, device, dtype, requires_grad, **kwar
     return inputs
 
 
+def sample_inputs_mixed_masked_reduction(op_info, device, dtype, requires_grad, **kwargs):
```
Instead of generating "mixed" sample inputs from OpInfos I would just model this as a generator specific to test. We can think of mixed layout testing as analogous to mixed dtype testing. OpInfo sample inputs typically don't test type promotion but we have specific tests which do. For example: 

https://github.com/pytorch/pytorch/blob/5e6f29661254e5ebc97fac3d829e5d455cda5864/test/test_binary_ufuncs.py#L566

The reason I'd suggest that approach is because OpInfos are meant to simplify the cross-product of what can be tested, so we don't generally want OpInfos to be worried about generating things like a cross-product of dtypes or layouts.

pearu committed:
Update on "Testing of masked reductions on mixed layout inputs."

The design of this feature is discussed in https://github.com/pytorch/pytorch/pull/71239#discussion_r787292751




[ghstack-poisoned]

pearu committed:
Update on "Testing of masked reductions on mixed layout inputs."

The design of this feature is discussed in https://github.com/pytorch/pytorch/pull/71239#discussion_r787292751




[ghstack-poisoned]

pearu committed:
Update on "Testing of masked reductions on mixed layout inputs."

The design of this feature is discussed in https://github.com/pytorch/pytorch/pull/71239#discussion_r787292751




[ghstack-poisoned]

pearu committed:
Update on "Testing of masked reductions on mixed layout inputs."

The design of this feature is discussed in https://github.com/pytorch/pytorch/pull/71239#discussion_r787292751




[ghstack-poisoned]

mruberry reviewed (approved):
Looks good from an OpInfo architecture perspective

cpuhrsch commented on test/test_masked.py:
```diff
@@ -162,10 +164,110 @@ def apply_masked_normalization_along_dim(op, input, *args, **kwargs):
+                        if layout != torch.sparse_coo and op.supports_sparse:
+                            sample_input_kwargs = sample_input.kwargs.copy()
+                            if mask.layout == torch.sparse_csr:
+                                # TODO: remove this if-block when sparse csr supports to_sparse
```
@pearu Could I ask you to send that PR separately as a follow up either as part of this stack or separately right after this lands?

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34408881).
